geo/src/algorithm/extremes.rs

```diff
@@ -1,274 +1,131 @@
-use super::kernels::*;
-use crate::prelude::*;
-use crate::*;
-use num_traits::Signed;
+use crate::algorithm::coords_iter::CoordsIter;
+use crate::{Coordinate, CoordinateType};
 
-// Useful direction vectors, aligned with x and y axes:
-// 1., 0. = largest x
-// 0., 1. = largest y
-// 0., -1. = smallest y
-// -1, 0. = smallest x
-
-/// Predicate that returns `true` if `vi` is (strictly) above `vj`
-/// along the direction of `u`
-fn above<T>(u: Coordinate<T>, vi: Coordinate<T>, vj: Coordinate<T>) -> bool
-where
-    T: CoordinateType + Signed + HasKernel,
-{
-    T::Ker::dot_product_sign(u, vi - vj) == Orientation::CounterClockwise
-}
-
-/// Predicate that returns `true` if `vi` is (strictly) below `vj`
-/// along the direction of `u`
-#[allow(dead_code)]
-fn below<T>(u: Coordinate<T>, vi: Coordinate<T>, vj: Coordinate<T>) -> bool
-where
-    T: CoordinateType + Signed + HasKernel,
-{
-    T::Ker::dot_product_sign(u, vi - vj) == Orientation::Clockwise
+/// Find the extreme coordinates and indices of a geometry.
+///
+/// # Examples
+///
+/// ```
+/// use geo::extremes::Extremes;
+/// use geo::polygon;
+///
+/// // a diamond shape
+/// let polygon = polygon![
+///     (x: 1.0, y: 0.0),
+///     (x: 2.0, y: 1.0),
+///     (x: 1.0, y: 2.0),
+///     (x: 0.0, y: 1.0),
+///     (x: 1.0, y: 0.0),
+/// ];
+///
+/// let extremes = polygon.extremes().unwrap();
+///
+/// assert_eq!(extremes.y_max.index, 2);
+/// assert_eq!(extremes.y_max.coord.x, 1.);
+/// assert_eq!(extremes.y_max.coord.y, 2.);
```

Comment:
It just occurred to me, but this is effectively the same thing as a bounding box, except that we also return the indices, right?

Reply:
@michaelkirk `Extremes` returns the most extreme coordinates of the geometry. In the case of a diamond, it will return each of the four coordinates. `BoundingRect` returns a minimum bounding `Rect` encompassing the geometry. The coordinates of the `Rect` are not necessarily vertices. In the case of a diamond, it will not return any of the four coordinates. The minimum bounding rectangle can be deduced from the extremes, but the reverse is not true. Does that answer your question?

Reply:
> In the case of a diamond, it will return each of the four coordinates

Ah, it returns the entire coord for each extreme, not just the extreme component of the coord. Thanks for explaining!

We could similarly use `exterior_coords` for bounding_rect. It seems like it's already doing everything optimally, so it'd just be a little code cleanup.

I think the only case where it would be less optimal is for Rects, for which we already know the min/max. You could apply a similar optimization for `Rect` in `extremes` to save yourself a couple comparisons if you're feeling like shaving off some nanograms, but seems like small potets.